airflow/models/dag.py

```diff
@@ -1373,6 +1374,40 @@ def normalized_schedule_interval(self) -> ScheduleInterval:
             _schedule_interval = self.schedule_interval
         return _schedule_interval
 
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def fetch_callback(
+        dag: DAG,
+        dagrun: DagRun | DagRunPydantic,
+        success: bool = True,
+        reason: str | None = None,
+        *,
+        session: Session = NEW_SESSION,
+    ) -> tuple[list[TaskStateChangeCallback], Context] | None:
+        """
+        Fetch the appropriate callbacks depending on the value of success.
+
+        This method gets the context of a single TaskInstance part of this DagRun and returns it along
+        the list of callbacks.
+
+        :param dag: DAG object
+        :param dagrun: DagRun object
+        :param success: Flag to specify if failure or success callback should be called
+        :param reason: Completion reason
+        :param session: Database session
+        """
+        callbacks = dag.on_success_callback if success else dag.on_failure_callback
+        if callbacks:
+            callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
+            tis = dagrun.get_task_instances(session=session)
+            ti = tis[-1]  # get first TaskInstance of DagRun
+            ti.task = dag.get_task(ti.task_id)
+            context = ti.get_template_context(session=session)
+            context["reason"] = reason
+            return callbacks, context
+        return None
+
     @provide_session
     def handle_callback(self, dagrun, success=True, reason=None, session=NEW_SESSION):
```

Comment:
```suggestion
    def handle_callback(self, dagrun: DagRun, success=True, reason=None, session=NEW_SESSION):
```

Reply:
Same here. When we add explicitly when we need  to use DagRun ORM version and when Pydantic version.